e2e/scenario_helpers_test.go

```diff
@@ -153,3 +159,73 @@ func getExpectedPackageVersions(packageName, distro, release string) []string {
 	}
 	return expectedVersions
 }
+
+func getCustomScriptExtensionStatus(ctx context.Context, t *testing.T, resourceGroupName, vmssName string) error {
+	pager := config.Azure.VMSSVM.NewListPager(resourceGroupName, vmssName, nil)
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get VMSS instances: %v", err)
+		}
+
+		for _, vmInstance := range page.Value {
+			instanceViewResp, err := config.Azure.VMSSVM.GetInstanceView(ctx, resourceGroupName, vmssName, *vmInstance.InstanceID, nil)
+			if err != nil {
+				t.Logf("failed to get instance view for VM %s: %v", *vmInstance.InstanceID, err)
+				continue
+			}
+			for _, extension := range instanceViewResp.Extensions {
+				for _, status := range extension.Statuses {
+					resp, err := parseLinuxCSEMessage(t, *status)
+					if err != nil {
+						return fmt.Errorf("Parse CSE message with error, error code: %s, raw message: %s", err.Code, *status.Message)
+					}
+					if !strings.EqualFold(resp.ExitCode, "0") {
+						return fmt.Errorf("vmssCSE %s, output=%s, error=%s", resp.ExitCode, resp.Output, resp.Error)
+					} else {
+						t.Logf("CSE completed successfully with exit code 0")
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func parseLinuxCSEMessage(t *testing.T, status armcompute.InstanceViewStatus) (*datamodel.CSEStatus, *datamodel.CSEStatusParsingError) {
```

Comment:
suggestion: verify with https://msazure.visualstudio.com/CloudNativeCompute/_git/aks-rp?path=/resourceprovider/server/microsoft.com/common/vmsshelper/csehelper.go&version=GBmaster&line=464&lineEnd=465&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents that we are covering all cases, nonetheless, good job getting this function out 🥇

Reply:
Yes, mostly copied the code from there and modified it a bit since AB e2es are much simpler, but we mostly care that the cse output is what we expect which this verifies.

Reply:
It's better to return `*datamodel.CSEStatus, error`

Using structs instead of `error` is causing a lot of issues. There is significant difference nil-interface and a nil-pointer. And at some stage one will be implicitly converted to another.

Reply:
For example, it may lead to situations like this:
```go
_, err1 := parseLinuxCSEMessage(...)
assert.Equal(err1, nil) // true
var err2 error
_, err2 := parseLinuxCSEMessage(...)
assert.Equal(err1, nil) // false
```